### PR TITLE
Feature: Add Chromecast support

### PR DESCRIPTION
**Pull request by tobiaseisenschenk**
I want to address #277 by adding support for Chromecast devices.

This feature implements [Google's design checklist](https://developers.google.com/cast/docs/design_checklist) by:
- add cast button in the `PlayerControlView` to discover devices
- sync queue and playback state to chromecast (e.g. play, seek, append etc.)
- generally uses the same UI as for local playback (mini- & expanded-player)
- mute local player when casting session is active

Overview of code changes:
- integrates custom Swift Package as the [GoogleCast framework](https://developers.google.com/cast/docs/ios_sender/integrate) does only support podfiles
- encapsulates casting code in `ChromecastPlayer`
- disables this feature for MacCatalyst (as the framework does not support it)

Tested on:
- iPhone 16 Simulator
- iPhone 11 Pro Device (Debug build)
- Mac (M2 Pro)

Open Questions:
- This project does not seem to use CocoaPods. The only other documented way to use this library is by static linking / adding the files manually. Is there perhaps a more elegant way?
- This the first time writing Swift for me - please suggest anything that needs work from you perspective!
- I might need to fix some edge cases such as re-ordering the context playlist, but wanted to get your feedback first.
- Is there a way I can test this as a release build or would I have to purchase the developer license?
- From the code I did not quite get what the purpose of the `userQueue` is in addition to the `contextQueue` and `activeQueue` is there any docs that explain the relation?

### CODE REVIEW

**Comment by tobiaseisenschenk** [comment]
On a related note, but probably not in the scope of this PR I wanted to raise the issue that with the added button the UI now looks very crowded. Maybe worth a discussion how this could be improved e.g. grouping airplay & chromecast? Let me know your thoughts!

Screenshot 2024-12-14 at 19 47 16

**Comment by Schlaubischlump** [comment]
Hi @tobiaseisenschenk,

I just checked the Cast framework documentation and it looks like MacCatalyst is supported, just not the static library. That could simplify the code quite a bit if I did not misunderstand the documentation. 

> Mac Catalyst setup
>
> For Apps that support Mac Catalyst, use the dynamic library of the Cast SDK. Follow the manual setup process to add the framework to your project. Then conditionally exclude the Cast SDK from the Mac target as mentioned in the Apple documentation. Static libraries are pre-compiled for iOS architecture, which causes a linker error when building against the Mac target

Have you looked into this?

**Comment by BLeeEZ** [comment]
Hi @tobiaseisenschenk,
first of all **thank you very much** for this PR. I can only imagine how much time this feature must have cost you.

> This project does not seem to use CocoaPods. The only other documented way to use this library is by static linking / adding the files manually. Is there perhaps a more elegant way?

I was using Carthage some time ago and moved away from it. Amperfy now is solely based on Swift Package Manager (SPM) which makes update/integration/build/release process much easier. From my point of view CocoaPods is outdated and I am not going to use it in this project.
I read the google documentation too and the only other official supported way is the static linking (which you used here). I found a topic on google bug tracker to support SPM in the future. Not much going on there ... (https://issuetracker.google.com/u/0/issues/190966385?pli=1). Due to licensing and update issues, copying the SDK file to this repository is not an option. I am not sure how we can fix that, since this is the biggest issue to submit this PR/feature.

> This the first time writing Swift for me - please suggest anything that needs work from you perspective!

I will need some time and give you feedback about my findings. As a first feedback: it looks really good for a first time Swift project :) 

> I might need to fix some edge cases such as re-ordering the context playlist, but wanted to get your feedback first.

Absolutely understandable. I am not sure how I can help here. I don't own a device with such capabilities.

> Is there a way I can test this as a release build or would I have to purchase the developer license?

There is some stuff you should be able to test without a developer license. You simply need to remove all capabilities from the entitlements file (Amperfy/Amperfy.entitlements). With that you can test Amperfy on your real devices (without CarPlay support ...)

> From the code I did not quite get what the purpose of the userQueue is in addition to the contextQueue and activeQueue is there any docs that explain the relation?

A detail explanation can be found here: https://github.com/BLeeEZ/amperfy/discussions/257

**Comment by tobiaseisenschenk** [comment]
Thanks for your quick and nice feedback!
I was able to remove the static linking of library files in this project by wrapping the lib in a [self-hosted swift package](https://github.com/tobiaseisenschenk/swift-package-google-cast-sdk), which is a much cleaner solution :)

If this is acceptable from your perspective I would
- have a look at the queue explanation you sent
- fix some edge cases
- look into thorough testing on my device

**Comment by BLeeEZ** [comment]
Hi @tobiaseisenschenk ,

moving all that dependencies in its own library makes it much easier to handle. Great approach.
I took a look at your code and here are a couple of remarks/questions:

- I don't own a Google Chrome Cast neither do I have experience with that functionality. This makes it difficult for me to say something about software architecture integration.
- Amperfy uses PlayQueueHandler as single point truth. This class knows exactly which item is playing/should be played, which item is next and so on. All my player classes are build around this class. The actual backend player gets the next item to be played inserted or requests the next item to load (in case of gapless). I don't know how that can be translated to the ChromCastPlayer. In the past I tried to keep the PlayQueueHandler and BackendPlayer in sync but this lead to a lot of problems. Your current approach tries to do that. ChromeCast has all songs from the past and future songs already inserted.
- In your current approach the BackendAudio player is only muted when ChromeCast is playing. This can result in a lot of problems, e.g. when the iOS user receives a phone call the playback is paused and continued when the call ends. This will be reflected in the Amperfy GUI, but the ChromeCast device will not be interrupted. In these cases the Amperfy GUI needs to decoupled from BackendPlayer and needs to be managed completely by ChromeCastPlayer.
- Can the user directly change the currently playing item on the target device. If that is true how should that be reflected in the Amperfy GUI and in the PlayQueueHandler?
- The playback can be changed via Siri void commands, Siri shortcuts X-Callback-URL and so on... can this be reflected in the ChromeCast too?
- I see that an ID is required for ChromeCast. Is this mandatory? Does it need to be requested at Google or can it be generated for each iOS device once or is one ID necessary for all Amperfy iOS installations?

I think that the biggest problem here is that I don't own a ChromeCast device and I am not able to maintain this feature in the future. The swift code itself looks good.

**Comment by tobiaseisenschenk** [comment]
Regarding the architecture:
I see your point that concerns are mixed between the `Chromecastplayer` and the `PlayQueueHandler`. This is done for three reasons:
- Initially had an issue starting a cast for the next track while the device is sleeping and sending the whole queue to the cast device solved it.
- There is some Chromecast specific features that rely on this technique. Two examples of this are that a casting device could display the list of tracks and there is a parameter to let the cast device download the upcoming track in order to ensure smooth playback.
- I also had a look at the Spotify implementation and it seems that they also built it such that the cast device is aware of the queue.

Regarding the described bugs:
- Thanks for pointing out these cases - I did not yet think of phone calls changing the playback state. I agree that handling this would need a decoupling, where the GUI reflects state of the ChromecastPlayer instead of the BackendAudioPlayer.
- I am not really familiar with the Siri integration (and might have had to disable it to develop locally). In general I'd say that incoming commands such as playback, volume, playback rate etc. could be handled and sent to the chromecast device.
- Are you referring to the `kGCKDefaultMediaReceiverApplicationID`? This is required by the library and is referencing a receiver application, which is not Amperfy, but the default app that runs on the casting device.

Regarding maintenance:
- Thank you a lot for the constructive and quick feedback!
- I completely understand that you can not merge something, that is not maintainable.
- I would be open to offer my support in maintaining this feature (e.g. fixing bugs, upgrading the dependency or checking a release). However this requires some coordination and could lead to slower releases as I can't guarantee my availability. Please let me know how your preference on this.

**Comment by BLeeEZ** [comment]
Ok, let’s continue with this feature. Here are some general ideas that come to my mind:

1. It must be possible to „easily“ remove this feature in cases of unmaintainability. I would suggest to place the main implementation inside of #if like we did for macOS.
2.  The chrome cast button should be moved to the (…) button menu.
3. Architecture wise I would suggest to have some kind of player proxy. The proxy decides which player (backend or chrome cast) is in charge.

**Comment by tobiaseisenschenk** [comment]
1. The whole feature is already in #if platform == ios else disabled (for the macos case) would you like to have an additional feature flag (e.g. `let chromeCastFeatureEnabled = true`), which would make it more clear what belongs to this feature or did you have something else in mind?
2. Ok, I will move it there
3. That's a good idea - I will try to refine and implement it.

Thanks for your willingness to add this to Amperfy @BLeeEZ !

**Comment by tobiaseisenschenk** [comment]
I think I need to pick your brain a bit regarding the proxy-player architecture @BLeeEZ :)
While implementing this pattern, it seemed to me that the cleanest solution would be to define the `ChromecastPlayer` as a second implementation of the `PlayerFacade` protocol. However the current implementation is declared as `lazy var  player` in both AppDelegate and its reference AmperKit. This means that assigning another value later (when the chromecast session starts or ends) has no effect, because lazy vars always return the initially computed value. Simply making this a computed + cached property is also not enough, because all views will be ignorant of changes. Therefore I tried to use the Combine Framework as in `@Published var playerObservable`, which seems doable for notifying VCs of a new player. This has some downsides though:
- the variable needs to have an assigned value, so we would be loosing the benefits of `lazy`
- all VCs, that reference `appDelegate.player` would need to define a sink and cancel on deinit

Therefore before making all these changes, I was wondering if you had any other setup in mind or if you think the above is a preferable option?

**Comment by BLeeEZ** [comment]
I was thinking about a similar approach like I already did for the backend API: https://github.com/BLeeEZ/amperfy/blob/master/AmperfyKit/Api/BackendProxy.swift.

With that approach all systems use the proxy and don't care which actual player is used. It is possible to switch between the `Chromecast` and `AVPlayer` via an additional method on the proxy.

**Comment by tobiaseisenschenk** [comment]
Great suggestion. I used the same approach now for a ProxyPlayer and this switching works. Next steps will be to
- Implement the PlayerFacade interface in the ChromecastAdapter
- Think about how the queue should work - I suppose we will need two one on the chromecast and one that holds Playables for the app to display. Also switching between local and casting playback should be seamless.

File: AmperfyKit/Player/BackendAudioPlayer.swift [hunk 1]
```diff
@@ -246,6 +263,7 @@ class BackendAudioPlayer: NSObject {
     
     func seek(toSecond: Double) {
         player.seek(to: CMTime(seconds: toSecond, preferredTimescale: CMTimeScale(NSEC_PER_SEC)))
+        onSeek!(toSecond)
```
[the remark this answers is not in the document]
Comment (cescofry) [reply AmperfyKit/Player/BackendAudioPlayer.swift:266]: suing ! will crash the app if onSeek is not provided.

File: AmperfyKit/Player/AudioPlayer.swift [hunk 2]
```diff
@@ -48,14 +48,16 @@ public class AudioPlayer: NSObject, BackendAudioPlayerNotifiable  {
     private let settings: PersistentStorage.Settings
     private let userStatistics: UserStatistics
     private var notifierList = [MusicPlayable]()
+    private var chromeCastPlayer: ChromecastPlayer
```
Comment (cescofry) [inline AmperfyKit/Player/AudioPlayer.swift:51]: since this is set dung init, it should be a let

File: AmperfyKit/Player/AudioPlayer.swift [hunk 3]
```diff
@@ -117,14 +119,17 @@ public class AudioPlayer: NSObject, BackendAudioPlayerNotifiable  {
             backendAudioPlayer.continuePlay()
             notifyItemStartedPlaying()
         }
+        self.chromeCastPlayer.playRemote()
```
Comment (cescofry) [inline AmperfyKit/Player/AudioPlayer.swift:122]: I am not sure I understand. So every time the user plays something we try to remote play on chromecast?
Same questions for all the other instances in this class.

File: Amperfy/Screens/Player/PopupPlayerVC.swift [hunk 4]
```diff
@@ -23,9 +23,12 @@ import UIKit
 import CoreMedia
 import AmperfyKit
 import PromiseKit
+import os.log
 
-class PopupPlayerVC: UIViewController, UIScrollViewDelegate {
+let kCastControlBarsAnimationDuration: TimeInterval = 0.20
```
Comment (cescofry) [inline Amperfy/Screens/Player/PopupPlayerVC.swift:28]: is this used anywhere?

File: Amperfy/Screens/Player/PopupPlayerVC.swift [hunk 5]
```diff
@@ -155,6 +162,21 @@ class PopupPlayerVC: UIViewController, UIScrollViewDelegate {
         tableView.reloadData()
         scrollToCurrentlyPlayingRow()
     }
+
+      func installViewController(_ viewController: UIViewController?, inContainerView containerView: UIView) {
```
Comment (cescofry) [inline Amperfy/Screens/Player/PopupPlayerVC.swift:166]: what part of the code calls these 2 functions?

File: Amperfy/Screens/Player/PlayerControlView.swift [hunk 6]
```diff
@@ -53,6 +56,14 @@ class PlayerControlView: UIView {
     @IBOutlet weak var displayPlaylistButton: UIButton!
     @IBOutlet weak var optionsButton: UIButton!
     
+    @IBOutlet var castInstructionLabel: UILabel!
+    @IBOutlet var credsToggleButton: UIButton!
+    @IBOutlet var credsLabel: UILabel!
```
Comment (cescofry) [inline Amperfy/Screens/Player/PlayerControlView.swift:61]: Where are these 2 used?

File: Amperfy/AppDelegate.swift [hunk 7]
```diff
@@ -26,6 +26,12 @@ import os.log
 import AmperfyKit
 import PromiseKit
 import Intents
+#if !targetEnvironment(macCatalyst)
+import GoogleCast
+#endif
+#if targetEnvironment(macCatalyst)
+let kGCKDefaultMediaReceiverApplicationID = ""
```
Comment (cescofry) [inline Amperfy/AppDelegate.swift:33]: No need to wrap in a `#if targetEnvironment(macCatalyst)`

File: Amperfy/AppDelegate.swift [hunk 8]
```diff
@@ -36,8 +42,10 @@ let defaultWindowActivityType = "amperfy.main"
 
 @UIApplicationMain
 class AppDelegate: UIResponder, UIApplicationDelegate {
-    
+     
     static let name = "Amperfy"
+    let kReceiverAppID = kGCKDefaultMediaReceiverApplicationID
+    let kDebugLoggingEnabled = true
```
Comment (cescofry) [inline Amperfy/AppDelegate.swift:48]: unused.

File: Amperfy/AppDelegate.swift [hunk 9]
```diff
@@ -232,6 +240,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         configureNotificationHandling()
         initEventLogger()
         self.window = UIWindow(frame: UIScreen.main.bounds)
+        #if !targetEnvironment(macCatalyst)
+        let criteria = GCKDiscoveryCriteria(applicationID: kReceiverAppID)
+        let options = GCKCastOptions(discoveryCriteria: criteria)
+            GCKCastContext.setSharedInstanceWith(options)
+        #endif
```
Comment (cescofry) [inline Amperfy/AppDelegate.swift:247]: I would extract all this code into let say a `ChromeCastContiguration` class and avoid cluttering the AppDelegate.

File: Amperfy.xcodeproj/project.pbxproj
Comment (cescofry) [inline Amperfy.xcodeproj/project.pbxproj:1]: Just discard the changes to the Project file.

File: Amperfy.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (cescofry) [inline Amperfy.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: Discard these changes too.
Also it looks like you added `swift-package-google-cast-sdk` which should generate changes to the Package file. But I do not see those changes.

**Review by cescofry — COMMENTED** [review]
This is mostly about style and Swift safety. I am not familiar enough with the architecture to comment on that.

**Comment by tobiaseisenschenk** [comment]
Hi,
Unfortunately I have decided to step away from working on this feature. From a technical perspective it requires much more conceptual thought than I anticipated. The difficulty is not in connecting to a device and streaming a track or playlist, but in keeping Amperfys serveral play queues synced with the queue on the casting receiver. Having two "source of truths" synced over network with different available metadata results in a bunch of non trivial edge cases. Those are not personally important to me, but should be solved when adding a feature to a production app.
The other reason is, that I own a Chromecast Ultra, which is a nowadays discontinued Google product and starts to become unreliable. I will therefore explore other alternatives to listen to my music on the living room speakers.

On that note I want to express my gratitude to everyone in this thread, who dedicated time to the discussion and review of my code. Thank you!
